File: ios/Packages/PadelCore/Sources/PadelCore/Model/PlayerProfile.swift
import Foundation

public enum Hand: String, Codable, CaseIterable, Sendable {
    case right
    case left

    public var wireName: String { rawValue }

    public static func fromWire(_ value: String) -> Hand {
        value == "left" ? .left : .right
    }
}

public struct PlayerProfile: Codable, Equatable, Sendable {
    /// Mano con la que se empuña la pala.
    public let hand: Hand
    /// Muñeca donde se lleva el reloj. Para medir golpeos **tiene que coincidir con
    /// `hand`**: el reloj solo ve el brazo en el que está.
    public let watchWrist: Hand
    public let birthYear: Int?
    public let maxHeartRate: Int?
    public let restingHeartRate: Int?

    public init(
        hand: Hand = .right,
        watchWrist: Hand = .right,
        birthYear: Int? = nil,
        maxHeartRate: Int? = nil,
        restingHeartRate: Int? = nil
    ) {
        self.hand = hand
        self.watchWrist = watchWrist
        self.birthYear = birthYear
        self.maxHeartRate = maxHeartRate
        self.restingHeartRate = restingHeartRate
    }

    /// Si es false, la UI debe avisar de que el conteo no será fiable.
    public var watchOnRacketArm: Bool { hand == watchWrist }

    /// FC máxima configurada, o estimada con 220 - edad. Sin datos, 190.
    public func effectiveMaxHeartRate(currentYear: Int) -> Int {
        if let maxHeartRate { return maxHeartRate }
        guard let birthYear else { return 190 }
        return min(max(220 - (currentYear - birthYear), 120), 210)
    }
}

File: android/core/src/main/kotlin/com/risingpadel/core/model/PlayerProfile.kt
package com.risingpadel.core.model

import kotlinx.serialization.Serializable

@Serializable
enum class Hand(val wireName: String) {
    RIGHT("right"),
    LEFT("left");

    companion object {
        fun fromWire(value: String): Hand = if (value == "left") LEFT else RIGHT
    }
}

/**
 * @param hand mano con la que se empuña la pala.
 * @param watchWrist muñeca donde se lleva el reloj. Para medir golpeos **tiene que
 *   coincidir con [hand]**: el reloj solo ve el brazo en el que está. Si no coinciden,
 *   [watchOnRacketArm] es false y la UI debe avisar de que el conteo no será fiable.
 * @param maxHeartRate FC máxima. Si es null se estima con 220 - edad a partir de
 *   [birthYear]; si tampoco hay año de nacimiento se usa 190.
 */
@Serializable
data class PlayerProfile(
    val hand: Hand = Hand.RIGHT,
    val watchWrist: Hand = Hand.RIGHT,
    val birthYear: Int? = null,
    val maxHeartRate: Int? = null,
    val restingHeartRate: Int? = null,
) {
    val watchOnRacketArm: Boolean get() = hand == watchWrist

    fun effectiveMaxHeartRate(currentYear: Int): Int =
        maxHeartRate ?: birthYear?.let { (220 - (currentYear - it)).coerceIn(120, 210) } ?: 190
}
